laddertools/ranking.py

```diff
@@ -111,7 +115,293 @@ def get_default_rating(cls):
+            if p0 is player:
+                return p1
+            elif p1 is player:
+                return p0
+            raise ValueError("Expected `player` to be involved in the game.")
+
+        def _get_outcome_for(game, player):
+            """Return 1 if `player` won the game, otherwise 0."""
+            if player_lookup[game.player0] is player:
+                return 1
+            elif player_lookup[game.player1] is player:
+                return 0
+            raise ValueError("Expected `player` to be involved in the game.")
+
+        start_date = min(map(lambda g: g.end_time, games))
+        start_date = start_date.replace(hour=0, minute=0, second=0)
+
+        games_by_period = _partition_games_in_rating_periods(
+            games, start_date, rating_period
+        )
+        current_registered_players = set()
+
+        for period, G in games_by_period.items():
+            # TODO: if no games in particular period, we should still increase the RD.
+            # Currently, we assume there's at least one game per period.
+            groups = _group_games_by_player(G, player_lookup)
+            current_registered_players = current_registered_players.union(groups.keys())
+            for player in current_registered_players:
+                G_ = groups.get(player, [])
+                outcomes = [_get_outcome_for(g, player) for g in G_]
+                opponents = [_get_opponent(g, player) for g in G_]
+                opponent_ratings = [
+                    _previous_period_rating_for(o, period) for o in opponents
+                ]
+                current_rating = _previous_period_rating_for(player, period)
+                period - rating_period
+                new_rating, tmp = RankingGlicko.compute_new_rating(
+                    current_rating,
+                    opponent_ratings,
+                    outcomes,
+                    return_intermediate_ratings=True,
+                )
+                for g, r in zip(G_, tmp):
+                    player_ratings_by_game.setdefault(g, [None, None])
+                    if player_lookup[g.player0] is player:
+                        assert player_ratings_by_game[g][0] is None
+                        player_ratings_by_game[g][0] = r
+                    elif player_lookup[g.player1] is player:
+                        assert player_ratings_by_game[g][1] is None
+                        player_ratings_by_game[g][1] = r
+                    else:
+                        raise ValueError
+
+                # We update the actual `r` value of `new_rating` by using the
+                # last intermediate rating. Strictly speaking this is not
+                # valid, because glicko-2 compute the rating once every
+                # `rating_period`. However, we want a new rating after every
+                # game, so we simply use the last intermediate rating as the
+                # 'official rating'. However, we do keep the RD and volatility
```

Comment:
2 However one after the other: are they excluding each other?

Reply:
Rephrased

Reply:
Official glicko-2 dictates the rating to be updated every (say) 2 weeks. That probably is not so great for openra where the players expect the rating to be updated after every game. So we compute an intermediate rating and return those intermediate ratings. But we only update the rating; we still use the official RD and volatility. As a consequence, the RD and volatility is constant over 2 weeks, and only after a new period has begun, the RD and volatility is changed (and it remains the same for another 2 weeks).